### PR TITLE
Refactor Tests Script for preparing ACVP python script Integration

### PR DESCRIPTION
**Pull request by potsrevennil**
Refactor `tests` script mainly for preparation to easily integrate acvp into `tests`:
- Split the original tests script into three files: the main `scripts/tests` script now acts as a Click-based CLI interface.
- Moved core testing logic to `scripts/lib/test.py`, providing APIs that the Click commands call for main testing functionality.
- Added scripts/lib/util.py for common type classes and utility functions.
- Rename the `State` class as mentioned in https://github.com/pq-code-package/mlkem-c-aarch64/pull/347#discussion_r1830477719)
- Make `tests func/kat/nistkat/all` interface consistent:
  - `--opt` flags become choice of `ALL | OPT | NO_OPT` (case insensitive)

### CODE REVIEW

**Review by mkannwischer — CHANGES_REQUESTED** [review]
Thanks @potsrevennil for the refactoring - this looks cleaner to me. 
I still need to go through the changes in detail, but maybe you can already have a look at this: 
`$ tests bench --opt all` does not work the way I would expect it - it only runs the opt tests.

File: .github/actions/functest/action.yml [hunk 1]
```diff
@@ -26,8 +26,8 @@ inputs:
     description: For auto-determining the default cross-prefix (e.g. native | cross )
     default: "native"
   opt:
-    description: opt flag to pass to test script
-    default: "true"
+    description: Determine for building opt/non-opt or both of the binaries (e.g. all | opt | no_opt)
```
[the remark this answers is not in the document]
Comment (potsrevennil) [reply .github/actions/functest/action.yml]: Thanks for the check, I forgot to run no-opt code when `tests bench --opt all`... 🤦‍♂️

[the remark this answers is not in the document]
Comment (potsrevennil) [reply .github/actions/functest/action.yml]: Note that when `--opt=all`, and if `--output <file>` is provided, the current tests script will benchmark both opt/no-opt implementations, but only write the output result to the output file.

Since we haven't yet decided how to output to the same file for both implementations. I've left a comment in the script about this, this should be sufficient for now.

**Review by mkannwischer — CHANGES_REQUESTED** [review]
Thanks for the updates @potsrevennil. 

~~This still doesn't do what I'd expect it to do:~~ nevermind - I was running an old version - let me try again.
```
$ tests bench --opt all --cycles PERF
None
INFO  > Benchmark       Compile     (native,    opt) make CROSS_PREFIX= bench CYCLES=PERF OPT=1 AUTO=1
INFO  > Benchmark       ML-KEM-512  (native,    opt)    keypair cycles = 24255
    encaps cycles = 30724
    decaps cycles = 39536

           percentile      1     10     20     30     40     50     60     70     80     90     99
   keypair percentiles:  23877  23975  24052  24114  24162  24255  24338  24450  24615  24919  26850
    encaps percentiles:  30305  30454  30519  30590  30652  30724  30815  30929  31134  31487  35613
    decaps percentiles:  39076  39228  39304  39393  39456  39536  39640  39727  39907  40208  42464

INFO  > Benchmark       ML-KEM-768  (native,    opt)    keypair cycles = 43647
    encaps cycles = 47745
    decaps cycles = 60480

           percentile      1     10     20     30     40     50     60     70     80     90     99
   keypair percentiles:  41991  42248  42475  42824  43240  43647  44009  44328  44817  45580  48292
    encaps percentiles:  45859  46158  46361  46649  47284  47745  48064  48347  48944  49807  52009
    decaps percentiles:  58633  58936  59164  59469  60010  60480  60869  61219  61821  62552  64655

INFO  > Benchmark       ML-KEM-1024 (native,    opt)    keypair cycles = 62950
    encaps cycles = 69488
    decaps cycles = 87193

           percentile      1     10     20     30     40     50     60     70     80     90     99
   keypair percentiles:  60969  61673  62105  62409  62658  62950  63238  63565  64200  64894  67050
    encaps percentiles:  67655  68272  68630  68972  69238  69488  69886  70193  70607  71357  73030
    decaps percentiles:  85259  85922  86317  86630  86906  87193  87540  87929  88307  89070  90809
```

**Review by mkannwischer — APPROVED** [review]
Thanks @potsrevennil - this looks good though. 
Hopefully the last rework of the entire test script. 

One downside is that we now no longer check the namespacing for the no-opt implementation in the quickcheck CI. Let's do that in a follow-up. I opened an issue #372
